File: app/models/questions/ClozeTestAnswer.java
/*
 * Copyright (c) 2018 The members of the EXAM Consortium (https://confluence.csc.fi/display/EXAM/Konsortio-organisaatio)
 *
 * Licensed under the EUPL, Version 1.1 or - as soon they will be approved by the European Commission - subsequent
 * versions of the EUPL (the "Licence");
 * You may not use this work except in compliance with the Licence.
 * You may obtain a copy of the Licence at:
 *
 * https://joinup.ec.europa.eu/software/page/eupl/licence-eupl
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the Licence is distributed
 * on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the Licence for the specific language governing permissions and limitations under the Licence.
 */

package models.questions;

import com.fasterxml.jackson.databind.JsonNode;
import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.Transient;
import java.lang.reflect.Type;
import java.util.Collections;
import java.util.Iterator;
import java.util.Map;
import java.util.Objects;
import java.util.regex.Pattern;
import models.base.GeneratedIdentityModel;
import models.sections.ExamSectionQuestion;
import org.apache.commons.lang3.math.NumberUtils;
import org.jsoup.Jsoup;
import org.jsoup.nodes.Attribute;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.jsoup.select.Elements;

@Entity
public class ClozeTestAnswer extends GeneratedIdentityModel {

    private static final String CLOZE_SELECTOR = "span[cloze=true]";

    private static final Pattern SPECIAL_REGEX_CHARS = Pattern.compile("[{}()\\[\\].+?^$\\\\/]");

    @Column(columnDefinition = "TEXT")
    private String answer;

    @Transient
    private String question;

    @Transient
    private Score score;

    public String getAnswer() {
        return answer;
    }

    public void setAnswer(String answer) {
        this.answer = answer;
    }

    public String getQuestion() {
        return question;
    }

    public Score getScore() {
        return score;
    }

    public ClozeTestAnswer copy() {
        ClozeTestAnswer clozeTestAnswer = new ClozeTestAnswer();
        clozeTestAnswer.setAnswer(answer);
        clozeTestAnswer.save();
        return clozeTestAnswer;
    }

    // This sets up the question so that it can be displayed to student
    public void setQuestion(ExamSectionQuestion esq) {
        Document doc = Jsoup.parse(esq.getQuestion().getQuestion());
        Elements blanks = doc.select(CLOZE_SELECTOR);
        blanks.forEach(b -> {
            boolean isNumeric = isNumeric(b);
            Iterator<Attribute> it = b.attributes().iterator();
            while (it.hasNext()) {
                Attribute a = it.next();
                if (!a.getKey().equals("id")) {
                    it.remove();
                }
            }
            b.tagName("input");
            b.text("");
            b.attr("aria-label", "cloze test answer");
            b.attr("type", isNumeric ? "number" : "text");
            b.attr("class", "cloze-input mt-2");
            if (isNumeric) {
                b.attr("step", "any");
                // Hacky, but this should allow for using both comma and period as decimal separator even in Firefox
                // regardless of browser language.
                b.attr("lang", "fi");
            }
        });
        this.question = doc.body().children().toString();
    }

    private void setQuestionWithResults(Document doc, String blankAnswerText, boolean showCorrect) {
        Map<String, String> answers = asMap(new Gson());
        Elements blanks = doc.select(CLOZE_SELECTOR);
        score = new Score();
        blanks.forEach(b -> {
            boolean isNumeric = isNumeric(b);
            String answer = answers.getOrDefault(b.attr("id"), "");
            boolean isCorrectAnswer = isCorrectAnswer(b, answer);
            String precision = b.attr("precision");
            if (isCorrectAnswer) {
                score.correctAnswers++;
            } else {
                score.incorrectAnswers++;
            }
            Iterator<Attribute> it = b.attributes().iterator();
            while (it.hasNext()) {
                it.next();
                it.remove();
            }
            b.text("");
            b.append(answer.isBlank() ? String.format("<em>%s</em>", blankAnswerText) : answer);
            if (showCorrect) {
                b.attr("class", isCorrectAnswer ? "cloze-correct" : "cloze-incorrect");
            } else {
                b.attr("class", "cloze-neutral");
            }
            if (isNumeric) {
                b.after("<span class=\"cloze-precision\">[&plusmn;" + precision + "]</span>");
            }
        });
        this.question = doc.body().children().toString();
    }

    // This sets up the question, so it can be displayed for review
    public void setQuestionWithResults(JsonNode esq, String blankAnswerText) {
        Document doc = Jsoup.parse(esq.get("question").get("question").asText());
        setQuestionWithResults(doc, blankAnswerText, true);
    }

    // This sets up the question, so it can be displayed for review
    public void setQuestionWithResults(ExamSectionQuestion esq, String blankAnswerText, boolean showCorrect) {
        Document doc = Jsoup.parse(esq.getQuestion().getQuestion());
        setQuestionWithResults(doc, blankAnswerText, showCorrect);
    }

    public Score calculateScore(ExamSectionQuestion esq) {
        Map<String, String> answers = asMap(new Gson());
        if (esq.getQuestion().getQuestion() == null) {
            return new Score();
        }
        Document doc = Jsoup.parse(esq.getQuestion().getQuestion());
        Elements blanks = doc.select(CLOZE_SELECTOR);
        Score score = new Score();
        blanks.forEach(b -> {
            String answer = answers.getOrDefault(b.attr("id"), "");
            boolean isCorrectAnswer = isCorrectAnswer(b, answer);
            if (isCorrectAnswer) {
                score.correctAnswers++;
            } else {
                score.incorrectAnswers++;
            }
        });
        return score;
    }

    private Map<String, String> asMap(Gson gson) {
        Type mapType = new TypeToken<Map<String, String>>() {}.getType();
        Map<String, String> map = gson.fromJson(answer, mapType);
        if (map != null) {
            map.values().removeIf(Objects::isNull);
            return map;
        }
        return Collections.emptyMap();
    }

    private boolean isNumeric(Element blank) {
        return Boolean.parseBoolean(blank.attr("numeric"));
    }

    private boolean isCorrectNumericAnswer(Element blank, String rawAnswer) {
        if (rawAnswer.isBlank()) {
            return false;
        }
        String answerText = rawAnswer.trim();
        if (!NumberUtils.isParsable(answerText)) {
            return false;
        }
        String precisionAttr = blank.attr("precision");
        double answer = Double.parseDouble(answerText);
        Double correctAnswer = Double.parseDouble(blank.text().trim().replaceAll("(^\\h*)|(\\h*$)", ""));
        Double precision = precisionAttr.isEmpty() ? 0.0 : Double.parseDouble(precisionAttr);
        return (correctAnswer - precision <= answer && answer <= correctAnswer + precision);
    }

    private String escapeSpecialRegexChars(String input) {
        return SPECIAL_REGEX_CHARS.matcher(input).replaceAll("\\\\$0");
    }

    private boolean isCorrectAnswer(Element blank, String rawAnswer) {
        if (isNumeric(blank)) {
            return isCorrectNumericAnswer(blank, rawAnswer);
        }
        // Get rid of excess whitespace
        String answer = rawAnswer.trim().replaceAll(" +", " ");
        String correctAnswer = blank
            .text()
            .trim()
            .replaceAll(" +", " ")
            .replaceAll(" \\|", "|")
            .replaceAll("\\| ", "|");
        // Generate the regex pattern. Replace '*' with '.*' and put the whole
        // thing in braces if there's a '|'.
        // For escaped '\*' and '\|' we have to first replace occurrences with special
        // escape sequence until restoring them in the regex.
        final String ESC = "__!ESC__";
        String regex = escapeSpecialRegexChars(correctAnswer)
            // Also backlashes will be escaped on escapeSpecialRegexChars call, therefore '\\*' pattern needs to be replaced
            .replaceAll("\\Q\\\\*\\E", ESC)
            .replace("*", ".*")
            .replace(ESC, "\\*")
            .replaceAll("\\Q\\\\|\\E", ESC);

        if (regex.contains("|")) {
            regex = String.format("(%s)", regex);
        }
        regex = regex.replace(ESC, "\\|");
        boolean isCaseSensitive = Boolean.parseBoolean(blank.attr("case-sensitive"));
        Pattern pattern = Pattern.compile(regex, isCaseSensitive ? 0 : Pattern.CASE_INSENSITIVE);
        return pattern.matcher(answer).matches();
    }

    // DTO
    public static class Score {

        int correctAnswers;
        int incorrectAnswers;

        public int getCorrectAnswers() {
            return correctAnswers;
        }

        public int getIncorrectAnswers() {
            return incorrectAnswers;
        }
    }
}

File: app/models/questions/ClozeTestAnswer.scala
// SPDX-FileCopyrightText: 2024 The members of the EXAM Consortium
//
// SPDX-License-Identifier: EUPL-1.2

package models.questions

import com.fasterxml.jackson.annotation.JsonAutoDetect
import com.fasterxml.jackson.annotation.JsonAutoDetect.Visibility
import com.fasterxml.jackson.annotation.JsonProperty
import com.fasterxml.jackson.databind.JsonNode
import com.google.gson.{Gson, reflect}
import jakarta.persistence.{Entity, Transient}
import models.base.GeneratedIdentityModel
import models.sections.ExamSectionQuestion
import org.apache.commons.lang3.math.NumberUtils
import org.jsoup.Jsoup
import org.jsoup.nodes.Element

import java.util.regex.Pattern
import scala.compiletime.uninitialized

@Entity
class ClozeTestAnswer extends GeneratedIdentityModel:
  @JsonProperty var answer: String = uninitialized

  @Transient var question: String             = uninitialized
  @Transient var score: ClozeTestAnswer.Score = uninitialized

  def copy(): ClozeTestAnswer =
    val c = new ClozeTestAnswer
    c.answer = answer
    c.save()
    c

  def setQuestion(esq: ExamSectionQuestion): Unit =
    val doc    = Jsoup.parse(esq.question.question)
    val blanks = doc.select(ClozeTestAnswer.ClozeSelector)
    blanks.forEach { b =>
      val isNumeric = ClozeTestAnswer.isNumericBlank(b)
      val id        = b.attr("id")
      b.clearAttributes()
      b.attr("id", id)
      b.tagName("input")
      b.text("")
      b.attr("aria-label", "cloze test answer")
      b.attr("type", if isNumeric then "number" else "text")
      b.attr("class", "cloze-input mt-2")
      if isNumeric then
        b.attr("step", "any")
        b.attr("lang", "fi")
    }
    question = doc.body().children().toString

  def setQuestionWithResults(esqNode: JsonNode, blankAnswerText: String): Unit =
    val doc = Jsoup.parse(esqNode.get("question").get("question").asText())
    doSetQuestionWithResults(doc, blankAnswerText, showCorrect = true)

  def setQuestionWithResults(
      esq: ExamSectionQuestion,
      blankAnswerText: String,
      showCorrect: Boolean
  ): Unit =
    val doc = Jsoup.parse(esq.question.question)
    doSetQuestionWithResults(doc, blankAnswerText, showCorrect)

  def calculateScore(esq: ExamSectionQuestion): ClozeTestAnswer.Score =
    val answers = asMap(new Gson)
    if esq.question.question == null then new ClozeTestAnswer.Score
    else
      val doc    = Jsoup.parse(esq.question.question)
      val blanks = doc.select(ClozeTestAnswer.ClozeSelector)
      val score  = new ClozeTestAnswer.Score
      blanks.forEach { blank =>
        val answer = answers.getOrElse(blank.attr("id"), "")
        if ClozeTestAnswer.isCorrectAnswer(blank, answer) then score.correctAnswers += 1
        else score.incorrectAnswers += 1
      }
      score

  private def doSetQuestionWithResults(
      doc: org.jsoup.nodes.Document,
      blankAnswerText: String,
      showCorrect: Boolean
  ): Unit =
    val answers = asMap(new Gson)
    val blanks  = doc.select(ClozeTestAnswer.ClozeSelector)
    score = new ClozeTestAnswer.Score
    blanks.forEach { b =>
      val isNumeric = ClozeTestAnswer.isNumericBlank(b)
      val answer    = answers.getOrElse(b.attr("id"), "")
      val isCorrect = ClozeTestAnswer.isCorrectAnswer(b, answer)
      val precision = b.attr("precision")
      if isCorrect then score.correctAnswers += 1 else score.incorrectAnswers += 1
      b.clearAttributes()
      b.text("")
      b.append(if answer.isBlank then s"<em>$blankAnswerText</em>" else answer)
      b.attr(
        "class",
        if showCorrect then if isCorrect then "cloze-correct" else "cloze-incorrect"
        else "cloze-neutral"
      )
      if isNumeric then b.after(s"<span class=\"cloze-precision\">[&plusmn;$precision]</span>")
    }
    question = doc.body().children().toString

  private def asMap(gson: Gson): Map[String, String] =
    import scala.jdk.CollectionConverters.*
    val mapType = new reflect.TypeToken[java.util.Map[String, String]]() {}.getType
    Option(gson.fromJson[java.util.Map[String, String]](answer, mapType))
      .fold(Map.empty[String, String])(_.asScala.filter((_, v) => v != null).toMap)

object ClozeTestAnswer:
  private val ClozeSelector     = "span[cloze=true]"
  private val SpecialRegexChars = Pattern.compile("[{}()\\[\\].+?^$\\\\/]")

  @JsonAutoDetect(fieldVisibility = Visibility.ANY)
  class Score:
    var correctAnswers: Int   = 0
    var incorrectAnswers: Int = 0

  private def isNumericBlank(blank: Element): Boolean =
    java.lang.Boolean.parseBoolean(blank.attr("numeric"))

  private def escapeSpecialRegexChars(input: String): String =
    SpecialRegexChars.matcher(input).replaceAll("\\\\$0")

  private def isCorrectNumericAnswer(blank: Element, rawAnswer: String): Boolean =
    if rawAnswer.isBlank then false
    else
      val answerText = rawAnswer.trim
      if !NumberUtils.isParsable(answerText) then false
      else
        val precisionAttr = blank.attr("precision")
        val answer        = answerText.toDouble
        val correctAnswer = blank.text().trim.replaceAll("(^\\h*)|(\\h*$)", "").toDouble
        val precision     = if precisionAttr.isEmpty then 0.0 else precisionAttr.toDouble
        correctAnswer - precision <= answer && answer <= correctAnswer + precision

  private def isCorrectAnswer(blank: Element, rawAnswer: String): Boolean =
    if isNumericBlank(blank) then isCorrectNumericAnswer(blank, rawAnswer)
    else
      val ESC    = "__!ESC__"
      val answer = rawAnswer.trim.replaceAll(" +", " ")
      val correctAnswer =
        blank.text().trim.replaceAll(" +", " ").replaceAll(" \\|", "|").replaceAll("\\| ", "|")
      var regex = escapeSpecialRegexChars(correctAnswer)
        .replaceAll("\\Q\\\\*\\E", ESC)
        .replace("*", ".*")
        .replace(ESC, "\\*")
        .replaceAll("\\Q\\\\|\\E", ESC)
      if regex.contains("|") then regex = s"($regex)"
      regex = regex.replace(ESC, "\\|")
      val isCaseSensitive = java.lang.Boolean.parseBoolean(blank.attr("case-sensitive"))
      val pattern = Pattern.compile(regex, if isCaseSensitive then 0 else Pattern.CASE_INSENSITIVE)
      pattern.matcher(answer).matches()
